### PR TITLE
Snforge utils and some common changes

### PR DESCRIPTION
**Pull request by immrsd**
Fixes [#1029](https://github.com/OpenZeppelin/cairo-contracts/issues/1029)

#### PR Checklist

- [x] Tests
- [ ] Documentation
- [ ] Added entry to CHANGELOG.md
- [ ] Tried the feature on a public network

### CODE REVIEW

File: src/tests/utils/events.cairo [hunk 1]
```diff
@@ -70,17 +72,17 @@ pub impl EventSpyExtImpl of EventSpyExt {
...
-    fn assert_no_events_left_from(ref self: EventSpy, contract_address: ContractAddress) {
+    fn assert_no_events_left_from(ref self: EventSpy, from_address: ContractAddress) {
         self.fetch_events();
-        assert!(self.count_events_from(contract_address) == 0, "Events remaining on queue");
+        assert!(self.count_events_from(from_address) == 0, "Events remaining on queue");
     }
 
-    fn count_events_from(self: @EventSpy, from: ContractAddress) -> u32 {
+    fn count_events_from(self: @EventSpy, from_address: ContractAddress) -> u32 {
         let mut result: u32 = 0;
```
Comment (ericnordelo) [inline src/tests/utils/events.cairo:81]:
```suggestion
        let mut result = 0;
```

File: src/tests/utils/events.cairo [hunk 2]
```diff
@@ -70,17 +72,17 @@ pub impl EventSpyExtImpl of EventSpyExt {
...
         let mut events = self.events.span();
-        while let Option::Some((from_address, _)) = events
+        while let Option::Some((from, _)) = events
             .pop_front() {
```
Comment (ericnordelo) [inline src/tests/utils/events.cairo:84]: While it doesn't matter in this context, popping for traversing an array is inefficient. We could something like do:

```
    fn count_events_from(self: @EventSpy, from_address: ContractAddress) -> u32 {
        let mut result = 0;
        let mut events = self.events.span();
        let mut remaining = events.len();

        while remaining > 0 {
            let (from, _) = events.at(remaining - 1);
            if from_address == *from {
                result += 1;
            }
            remaining -= 1;
        };
        result
    }
```

Comment (ericnordelo) [reply src/tests/utils/events.cairo:84]: We do have iterator syntax in the next Cairo release, which should make this easier.

File: src/tests/utils/events.cairo [hunk 3]
```diff
@@ -0,0 +1,91 @@
+use core::array::ArrayTrait;
+use core::array::SpanTrait;
+use openzeppelin::upgrades::UpgradeableComponent::Upgraded;
+use openzeppelin::upgrades::UpgradeableComponent;
```
Comment (ericnordelo) [inline src/tests/utils/events.cairo:4]:
```suggestion
```

File: src/tests/utils/events.cairo [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+use snforge_std::cheatcodes::events::EventFetcher;
+use snforge_std::{spy_events, SpyOn, Event, EventSpy, EventAssertions};
```
Comment (ericnordelo) [inline src/tests/utils/events.cairo:6]:
```suggestion
use snforge_std::{spy_events, SpyOn, EventSpy, EventAssertions};
```

File: src/tests/utils/events.cairo [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+use starknet::{ContractAddress, ClassHash};
```
Comment (ericnordelo) [inline src/tests/utils/events.cairo:7]:
```suggestion
use starknet::ContractAddress;
```

File: src/tests/token.cairo [hunk 6]
```diff
@@ -1,3 +1,5 @@
 // pub(crate) mod erc1155;
 pub(crate) mod erc20;
 // pub(crate) mod erc721;
+
+
```
Comment (ericnordelo) [inline src/tests/token.cairo:5]:
```suggestion
```

File: src/tests/utils.cairo [hunk 7]
```diff
@@ -1,40 +1,41 @@
...
 pub(crate) mod constants;
-pub mod foundry;
-pub use foundry::declare_and_deploy;
+pub(crate) mod events;
+pub(crate) mod signing;
+pub use common::{declare_class, declare_and_deploy, declare_and_deploy_at, deploy, deploy_at};
 
-pub use foundry::spy_on;
+pub use events::{spy_on, EventSpyExt};
+// use starknet::ContractAddress;
```
Comment (ericnordelo) [inline src/tests/utils.cairo:8]: Why are these commented? We can remove the commented functions.

**Review by ericnordelo — COMMENTED** [review]
Looking good @immrsd. I left some comments.

File: src/tests/utils.cairo (hunk 7 above)
Comment (immrsd) [reply src/tests/utils.cairo]: Removed

File: src/tests/utils/events.cairo [hunk 8]
```diff
@@ -13,16 +13,20 @@ pub fn spy_on(contract_address: ContractAddress) -> EventSpy {
 #[generate_trait]
 pub impl EventSpyExtImpl of EventSpyExt {
     fn assert_only_event<T, impl TEvent: starknet::Event<T>, impl TDrop: Drop<T>>(
```
[the remark this answers is not in the document]
Comment (immrsd) [reply src/tests/utils/events.cairo]: Fixed

File: src/tests/utils/events.cairo [hunk 9]
```diff
@@ -13,16 +13,20 @@ pub fn spy_on(contract_address: ContractAddress) -> EventSpy {
...
-        ref self: EventSpy, from: ContractAddress, event: T
+        ref self: EventSpy, from_address: ContractAddress, event: T
     ) {
-        self.assert_emitted_single(from, event);
-        self.assert_no_events_left_from(from);
+        self.assert_emitted_single(from_address, event);
+        self.assert_no_events_left_from(from_address);
     }
 
     fn assert_emitted_single<T, impl TEvent: starknet::Event<T>, impl TDrop: Drop<T>>(
```
[the remark this answers is not in the document]
Comment (immrsd) [reply src/tests/utils/events.cairo]: Fixed

File: src/tests/utils/events.cairo [hunk 10]
```diff
@@ -43,16 +47,14 @@ pub impl EventSpyExtImpl of EventSpyExt {
         self.events = new_events;
     }
 
-    fn drop_events_from(
-        ref self: EventSpy, contract_address: ContractAddress, number_to_drop: u32
-    ) {
+    fn drop_events_from(ref self: EventSpy, from_address: ContractAddress, number_to_drop: u32) {
         self.fetch_events();
         let mut dropped_number: u32 = 0;
```
[the remark this answers is not in the document]
Comment (immrsd) [reply src/tests/utils/events.cairo]: Fixed

File: src/tests/utils/events.cairo (hunk 4 above)
Comment (immrsd) [reply src/tests/utils/events.cairo:2]: We need `Event` in **drop_events_from** and **drop_all_events_from** functions

File: src/tests/utils/events.cairo [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    fn drop_events_from(ref self: EventSpy, from_address: ContractAddress, number_to_drop: u32) {
+        self.fetch_events();
+        let mut dropped_number = 0;
+        let mut new_events: Array<(ContractAddress, Event)> = array![];
+        while let Option::Some((from, event)) = self
+            .events
+            .pop_front() {
+                if from == from_address && dropped_number < number_to_drop {
```
Comment (andrew-fleming) [inline src/tests/utils/events.cairo:53]: I think it makes sense to define this right under `drop_event_from` to make this module (and eventually documentation) even tidier. Part of me also wonders if this should be renamed `drop_n_events_from` (and reverse the parameter order). WDYT?

File: src/tests/utils/events.cairo [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    fn assert_no_events_left_from(ref self: EventSpy, from_address: ContractAddress) {
+        self.fetch_events();
+        assert!(self.count_events_from(from_address) == 0, "Events remaining on queue");
+    }
+
+    fn count_events_from(self: @EventSpy, from_address: ContractAddress) -> u32 {
+        let mut result = 0;
+        let mut events = self.events.span();
```
Comment (andrew-fleming) [inline src/tests/utils/events.cairo:78]: Do you think we should have `count_events` as well to be comprehensive?

**Review by andrew-fleming — COMMENTED** [review]
Nice work on this! I left a few questions

File: src/tests/utils/events.cairo (hunk 12 above)
Comment (immrsd) [reply src/tests/utils/events.cairo:78]: It's hard to think of use cases for `count_events` function. For now, I don't want to add it, but if there's a need for it we can surely do it

File: src/tests/utils/common.cairo [hunk 13]
```diff
@@ -0,0 +1,36 @@
+use core::poseidon::{Poseidon, PoseidonTrait};
+use core::traits::TryInto;
```
[the remark this answers is not in the document]
Comment (immrsd) [reply src/tests/utils/common.cairo:2]: Yeah, removed it

File: src/tests/utils/events.cairo (hunk 11 above)
Comment (immrsd) [reply src/tests/utils/events.cairo:53]: I like the `drop_n_events_from` name for the function, it makes the difference to the `drop_event_from` more explicit. I've also moved the function to place them next to each other

Regarding reversing the parameter order, I reckon it's better to leave it as it is. The `from_address` param is the dominant one here, and moreover that keeps parameter-order approach consistent within the module

Comment (andrew-fleming) [reply src/tests/utils/events.cairo]: > Regarding reversing the parameter order, I reckon it's better to leave it as it is. The from_address param is the dominant one here, and moreover that keeps parameter-order approach consistent within the module

Fair points, sounds good

**Review by andrew-fleming — APPROVED** [review]
LGTM!

**Review by ericnordelo — APPROVED** [review]
Nice job! LGTM!

**Pull request merged**